kernel/copy.masm

```diff
@@ -0,0 +1,256 @@
+  dup.13 neq.0 while.true
+    # stack: [A', B', C', writePtr', words, ...]
+    adv_pipe
+
+    # move `words` counter up
+    # stack: [[writePtr', words, x, x], A', B', C', ...]
+    swapw.3
+
+    # stack: [[words, writePtr', x, x], A', B', C', ...]
+    movup.1
+
+    # optimization: `words` is required to be even and greater-than zero for
+    # this procedure to be used. Here it is possible to do a division instead
+    # of a range check, so the test is a single cycle
+    # stack: [[words/2, writePtr', x, x], A', B', C', ...]
+    div.2
+
+    # move `words` counter down
+    # LOOP: [A', B', C', writePtr', words/2, ...]
+    movdn.1 swapw.3
+
+    # while(words>0)
+    dup.13 neq.0
+  end
+
+  # stack: [A', B', C', writePtr', ...]
+  movup.13 drop
+end
+
+# move_from_tape_to_memory(words: Felt, writePtr: Address) -> [RpoHash, writePtr']
+#
+# Moves the `words` words from the advice tape to memory, starting at
+# `writePtr` address, the hash of the copied data is left on the stack for
+# commitment checking.
+proc.move_from_tape_to_memory.0
+  # input: [words, writePtr, ...]
+  # output: [A, B, C, writePtr', ...]
+  
+  # check if there is an odd number of words
+  # stack: [is_odd, words, writePtr, ...]
+  dup exec.is_odd
+
+  # copy the flag `is_odd`, this will later on define if padding is required
+  # stack: [is_odd, words, writePtr, needs_padding, ...]
+  dup movdn.3
+
+  # Make sure `words` is event. This is a requirement for the
+  # `move_from_tape_to_memory_even_words`.
+  # - even words will be `words-0` so result stays even
+  # - odd words will be `words-1` so result becomes event
+  #
+  # stack: [words_even, writePtr, needs_padding, ...]
+  sub
+
+  # Prepare the capacity word. For rescue prime optimized the first element is
+  # set to `1` when padding is used and `0` otherwse, this is determined by the
+  # `needs_padding` flag, so just copy it.
+  dup.1 push.0.0.0
+
+  # set initial state
+  # stack: [A, B, C, writePtr, words_even, needs_padding, ...]
+  padw padw
+
+  # stack: [A, B, C, writePtr, needs_padding, ...]
+  exec.move_from_tape_to_memory_even_words
+
+  # stack: [needs_padding, A, B, C, writePtr, ...]
+  movup.13
+
+  # if(needs_padding)
+  if.true 
+    # Rescue Prime Optimized uses overwrite mode. Drop `A`.
+    # stack: [B, C, writePtr, ...]
+    dropw
+
+    # Overwrite the `B` with the new data
+    # stack: [B', C, writePtr, ...]
+    adv_loadw
+
+    # - get the memory address that B' should be saved to
+    # - update the writePtr to point to the next address
+    # stack: [writePtr, B', writePtr+1, C, ...]
+    movup.8 dup.0 add.1 movdn.5
+
+    # save data to memory
+    # stack: [B', writePtr+1, C, ...]
+    mem_storew
+
+    # Fix writePtr position
+    # stack: [B', C, writePtr+1, ...]
+    movup.4 movdn.8
+
+    # Push padding word
+    # stack: [A, B', C, writePtr+1, ...]
+    push.1.0.0.0
+
+    # Run RPO permutation
+    # stack: [A', B', C', writePtr+1, ...]
+    rpperm
+  end
```

Comment:
I'll need to think a bit more about how we want to handle padding here. It might be simpler to require the padding be done correctly outside of the VM rather than doing in the VM (i.e., the advice provider will contain padded data already). I don't believe this will be a security risk - but need to think it through.

Another thing: currently we are specifying the number of words to read, but shouldn't it be the number of elements? Or if we wanted to make this limited to words, maybe the name of the procedure should be `move_word_from_tape_to_memory`?

Reply:
> It might be simpler to require the padding be done correctly outside of the VM rather than doing in the VM

No problem, the code should handle both cases. At least for the moment I was optimizing for flexibility, that is why this is built-in. I think once we have the data structures properly defined we can write custom code that doesn't use loops to load the data.

> currently we are specifying the number of words to read, but shouldn't it be the number of elements?

I wrote the procedures with words because that is what we have on the issue spec. I can rename it and add another version that works on elements too.